### ISSUE TITLE
create some sort of proxy for problems on non-standard ports

### ISSUE THREAD

**Issue by carlislemc** [picoCTF/picoCTF#65]
Many schools seem to block these ports. If we had a proxy that went to them (and not anywhere else on the internet!), that might help.

**Comment by hiliang-cmu** [comment]
Does the availability of the webshell address the problem of access? 

Students in a restricted environment might not be able to configure proxies. Another solution could be along lines of a simple nginx `proxy_pass` that redirects `https://proxy.picoctf.com/<port>/<requestURL>` to the shell server(s).

**Comment by carlislemc** [comment]
The webshell doesn't really address this. If we are hosting a web challenge on port 50000, they can't use a web browser from the shell. Something like the proxy_pass you suggest would be a nice enhancement.

**Comment by hiliang-cmu** [comment]
The presence of multi shell-servers makes me think this proxy URL would need to be in every problem description. Given something like:

`There is a website running at http://{{server}}:{{port}}. Try to see if you can login!`,

turns into:

`There is a website running at http://{{server}}:{{port}} or http://{{proxyserver}}/{{port}}. Try to see if you can login!`,

We might as well make the 'proxy' the default one displayed.

**Comment by carlislemc** [comment]
I agree.

**Comment by hiliang-cmu** [comment]
In attempting to apply the aforementioned proxy scheme on one of the test servers, two problems arise that make this unlikely to sort out in short term:

1. If the port, say `1234` is specified in the request path (e.g. `https://shell.picoctf.com/1234/) then the pages served will not be sitting at root path, and absolute links such as `/login` will wind up breaking. To avoid this would require problem writers to make all resources (css, js, images) as well as links to use relative paths. 

2. Trying to be clever with a subdomain, e.g. `https://1234.shell.picoctf.com/` would address the path issue, but this doesn't work with our wildcard SSL cert. The one our registrar currently provides works only for 1 level subdomain wildcard, `*.picoctf.com`. It's not as simple as adding a CNAME to get a new second level wildcard domain cert for `*.proxy.picoctf.com`. If this were possible, we could have possibly something like `https://2019shell2-1234.proxy.picoctf.com`.

**Comment by maverickwoo** [comment]
For brainstorming...

Regarding idea 1, would the following pair of mod_rewrite rules work? (I didn't think through this too much and I assume nginx has the same capability in rewriting as apache.)

(a) https://shell.picoctf.com/1234/ internal proxies (P) to https://shell.picoctf.com:1234/. (Obviously, we can't use an external redirect and so we need to use P.)

(b) https://shell.picoctf.com/xyz with a referrer of https://shell.picoctf.com/1234/ proxies to https://shell.picoctf.com:1234/xyz. (Since this depends on the referrer field, this still assumes the problem text does not tell the user to access a URL directly, i.e., not through a hyperlink.)

**Comment by hiliang-cmu** [comment]
Almost...I added step b into the mix. Problem is any subsequent action loses the referral.

Very specific example from problem `Logon`:

`https://shell.picoctf.com/1234/` has a link to `/logout`. 

This is rewritten+proxied properly by part b to `https://shell.picoctf.com/1234/logout`.  

Now to the user's browser, the url has become `https://shell.picoctf.com/logout` which means subsequent `referrer` field has changed and loses all info on its original port origin.

**Comment by maverickwoo** [comment]
Hmm, how about in (b) we do an external redirect (R in mod_rewrite) so that the URL at the browser becomes `https://shell.picoctf.com/1234/xyz`?

**Comment by hiliang-cmu** [comment]
I think I got it, I had quickly given up on redirect due to what I thought was the referrer getting lost in my nginx directives. Turns out it was because the problem's own flask redirect function was setting `Location` to HTTP, which made chrome's default `Referrer-Policy: no-referrer-when-downgrade` lose the referrer.

Setting `proxy_redirect http:// $scheme://;` that intercepts those Location type of responses and retains the referrer throughout.

Only remaining thing that I can see, is to make sure problem writers reference external css/js with HTTPS urls.
